test/studies/ml/lib/Chai.chpl

```diff
@@ -0,0 +1,760 @@
+        var padding: int = 0;
+
+        proc init(inChannels: int,outChannels: int, kernelSize: int = 3, stride: int = 1, padding: int = 0) {
+            const numFilters = outChannels;
+            this.numFilters = numFilters;
+
+            const fanIn = kernelSize * kernelSize * inChannels;
+            const fanOut = outChannels;
+            const stddev = sqrt(2.0 / (fanOut + fanIn));
+            this.filters = tn.randn(outChannels,kernelSize,kernelSize,inChannels,mu=0.0,sigma=stddev);
+            // this.filters = tn.randn(numFilters,kernelSize,kernelSize,inChannels) / (kernelSize:real ** 2.0);
+            
+            this.filtersGrad = tn.zeros(numFilters,kernelSize,kernelSize,inChannels);
+            this.stride = stride;
+            this.padding = padding;
+        }
+
+        proc forwardPropBatch(batch: [] Tensor(3)): [] Tensor(3) {
+            const batchSize = batch.size;
+            var convs: [0..#batchSize] Tensor(3);
+            forall (image,i) in zip(batch,0..) {
+                convs[i] = forwardProp(image);
+            }
+            return convs;
+        }
+
+        proc forwardProp(images: Tensor(3)): Tensor(3) {
+            const (h,w,channels) = images.shape;
+            const (outChannels,kh,kw,inChannels) = filters.shape;
+
+            if channels != inChannels {
+                writeln("input: ", images.shape);
+                writeln("filters: ", filters.shape);
+                err("Conv forwardProp: inChannels mismatch");
+            }
+
+            var convs = new Tensor(3,real);
+            const (newH,newW) = correlateShape((kh,kw),(h,w),stride,padding);
+            convs.reshapeDomain({0..#newH, 0..#newW, 0..#outChannels});
+            forall f in 0..#outChannels with (ref convs, var filter: Tensor(3) = tn.zeros(kh,kw,channels)) {
+                filter.data = filters[f,..,..,..];
+                convs[..,..,f] = correlate(filter=filter,input=images,stride=stride,padding=padding);
+            }
+            convs.data /= (inChannels:real);
+            return convs;
+        }
+
+        proc backward(delta: Tensor(3), images: Tensor(3)): Tensor(3) {
+            const (h,w,channels) = images.shape;
+            const (outChannels,kh,kw,inChannels) = filters.shape;
+            const (dh,dw,dc) = delta.shape;
+
+
+            if dc != outChannels then err("Conv backward: outChannels mismatch");
+            if channels != inChannels then err("Conv backward: inChannels mismatch");
+
+            const dL_dF = tn.filterGradient(images,delta,stride,padding,kh);
+            filtersGrad += dL_dF;
+
+            var dL_dX = new Tensor(3,real);
+            dL_dX.reshapeDomain({0..#h, 0..#w, 0..#inChannels});
+            forall (m,n,ci) in {0..#h, 0..#w, 0..#inChannels} with (ref dL_dX) {
+                var sum = 0.0;
+                forall co in 0..#outChannels with (+ reduce sum) {
+                    forall (i,j) in {0..#dh, 0..#dw} with (+ reduce sum) {
```

Comment:
Is the reason for `forall`s so that you can use reduce intents? I expect `for` to perform a bit better in the nested loops here.

Reply:
`{0..#dh, 0..#dw}` is an invariant domain. Could you put this outside of this deep loop nest? Domain creation is not cheap and you're paying for it in the tightly-nested inner loop here.

Reply:
Which of the forall loops should I pick to be serial? Usually, `outChannels` will be between `1-200`, but `(dh,dw,dc) = delta.shape` will be the dimensions of an image. For example, a 500x500 pixel RGB image would be `(500,500,3)`. The innermost computation is relatively small.